source/extensions/filters/http/adaptive_concurrency_limit/limit/gradient.cc

```diff
@@ -0,0 +1,147 @@
+#include "extensions/filters/http/adaptive_concurrency_limit/limit/gradient.h"
+
+#include <cmath>
+
+#include "envoy/registry/registry.h"
+
+#include "common/access_log/access_log_formatter.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace AdaptiveConcurrencyLimit {
+namespace Limit {
+
+Gradient::Gradient(
+    const envoy::config::filter::http::adaptive_concurrency_limit::v2alpha::
+        AdaptiveConcurrencyLimit::Limit::CommonConfig& common_config,
+    const envoy::config::filter::http::adaptive_concurrency_limit::v2alpha::GradientLimitConfig&
+        limit_specific_config,
+    Runtime::RandomGenerator& random, const std::string& cluster_name)
+    : random_(random), cluster_name_(cluster_name),
+      min_limit_(PROTOBUF_GET_WRAPPED_REQUIRED(common_config, min_limit)),
+      max_limit_(PROTOBUF_GET_WRAPPED_REQUIRED(common_config, max_limit)),
+      smoothing_(PROTOBUF_GET_WRAPPED_REQUIRED(limit_specific_config, smoothing)),
+      rtt_tolerance_(PROTOBUF_GET_WRAPPED_REQUIRED(limit_specific_config, rtt_tolerance)),
+      probe_interval_(limit_specific_config.has_probe_interval()
+                          ? absl::optional<uint32_t>{limit_specific_config.probe_interval().value()}
+                          : absl::nullopt),
+      probe_countdown_(nextProbeCountdown()),
+      estimated_limit_((PROTOBUF_GET_WRAPPED_REQUIRED(common_config, initial_limit))) {}
+
+double Gradient::getQueueSize(double estimated_limit) {
+  // The square root of the limit is used to get queue size
+  // because it is better than a fixed queue size that becomes too
+  // small for large limits. Moreover, it prevents the limit from growing
+  // too much by slowing down growth as the limit grows.
+  return sqrt(estimated_limit);
```

Comment:
Sqrt can be an expensive operation.  Not sure if c++ optimizes square roots for small numbers but it might be a good idea to implement a simple lookup table for limits under 1000.